### DISCUSSION TITLE
I thought react-spring (useSpring) causes the component to re-render, but it may not?

### DISCUSSION THREAD

**Discussion by nonopolarity** [pmndrs/react-spring#1436] [Support] [answered]
## 🤓 Question

I guess the main question is, how do we know what situation will react-spring re-render a lot and how to we prevent it if possible?

https://stackoverflow.com/questions/67188777/i-thought-react-spring-usespring-causes-the-component-to-re-render-but-it-may

I thought react-spring `useSpring()` causes the component to re-render a lot, so if it is a component that already has a lot of CPU intensively work to do, then react-spring is not best suited for the task.

However, I can see that it re-renders a lot, in:

https://codesandbox.io/s/musing-dew-9tfi9?file=/src/App.tsx

(by looking at the console.log output, which has a lot of print out of `renderCount`.  The print out is a lot more when we change the duration to `5000` which is 5 seconds).

Likewise, if it is a component that is similar to react-spring, it'd render a lot:

https://codesandbox.io/s/wonderful-currying-9pheq

However, the following code:

```jsx
let renderCount = 0

export default function App() {
  const styles = useSpring({
    loop: true,
    to: [
      { opacity: 1, color: '#ffaaee' },
      { opacity: 0.5, color: 'rgb(14,26,19)' },
      { transform: 'translateX(100px)' },
      { transform: 'translateX(0px)' },
    ],
    from: { opacity: 0, color: 'red', transform: 'translateX(0px)' },
    config: { duration: 500 },
  })

  console.log('renderCount', ++renderCount)

  return <a.div style={styles}>I will fade in and out</a.div>
}
```
Demo: https://codesandbox.io/s/dazzling-rgb-j2bx3?file=/src/App.tsx

We can see that the `renderCount` hardly get printed out at all.  `react-spring` should need to keep on updating the `style` of the component, so after a minute, I'd expect a lot of print out of `renderCount` like the first two examples above, but it does not.

How and why doesn't react-spring cause a lot of re-rendering in this case, and how do we know in what situation would react-spring cause a lot of re-rendering (and how to prevent it)?

**Comment by joshuaellis** [comment, marked as answer]
The API allows spring to animate outside of the react-render loop. In the first example you shared, you're using state to dictate the props of the spring. In the third example you've shared you've got your spring defined and ready so the `animated` component does not need to have a re-render to have it's styles updated.

You could also look at using the [Imperative API](https://react-spring.io/common/imperatives-and-refs) we now have which would allow you to change the state of the spring in reaction to events without updating the state of the component. So that's the kind of how and why but in summary:

HOW – the animated component
WHY – it's better and more performance efficient to be honest.

How to prevent it? IMO use the imperative api, i linked above, this is how I write all my springs and I haven't had any issues so far.

**Reply by nonopolarity** [reply]
> How to prevent it? IMO use the imperative api, i linked above, this is how I write all my springs and I haven't had any issues so far.

Thank you for your explanation of the difference.

So imperative is the one that will cause a lot of re-render?

Or is imperative the one that doesn't cause a lot of re-render?  And are you saying you prefer and use the one that doesn't cause a lot of re-render yourself all the time?

Usually, it doesn't matter to me, but I am working on a project that the side panel is already too busy rendering all the data constantly (in a loop, say, every 0.3 seconds), and so is the main page... so the idea is to "click on the Update Main Page button" and then this side panel will slide away of the main page and then stop updating itself.  So if I use react-spring to "imperative" slide it out, it is going to cause more issue because it cannot afford to get busier... it will slow the page even more (during the transition of sliding it out)... luckily only for 0.5 seconds only, but if it is choppy animation, it wouldn't be as nice as a smooth animation.

P.S. I tried the imperative API which you linked earlier:
https://codesandbox.io/s/reverent-oskar-0yyxb
so it doesn't have repeated rendering... you mean this is the way you write all your springs?

Yes, I did get the "deprecation notice"... so... how should I write it using the current version?  Can you give 2 codesandbox to show the difference?  But codesandbox can be altered for the code by the original author, so I find it quite troublesome. (perhaps one way is to use two different accounts... one for trying and editing, and one for "forking and freezing" a snapshot for linking on your website)

**Reply by joshuaellis** [reply]
> Or is imperative the one that doesn't cause a lot of re-render? And are you saying you prefer and use the one that doesn't cause a lot of re-render yourself all the time?

This is what I mean.

And yes, I write all my springs this way, I much prefer the control I have, I can use that API to react to events & use in effects etc. Make sure you check out the API for it, it's more powerful than you think – https://react-spring.io/common/imperatives-and-refs#api-methods

**Reply by nonopolarity** [reply]
are the docs also on the same github?  the word "depreceation" need to be changed but I don't see the docs on github to make a PR.

**Reply by joshuaellis** [reply]
check out [pmndrs/react-spring.io](https://www.github.com/pmndrs/react-spring.io)

**Reply by nonopolarity** [reply]
so you mean just change `api` to `api.start`?  https://codesandbox.io/s/zen-rain-066ly?file=/src/App.js

**Reply by joshuaellis** [reply]
For your example yes. But there are more methods which can be useful

**Reply by nonopolarity** [reply]
interesting... I made an interactive example: https://codesandbox.io/s/wizardly-leaf-814y9
but maybe your readers are advanced enough to know how to do it interactively...
but does react-spring have something similar to   `transform: translateX(-100%)` to make itself disappear by moving to the left for its whole width? (instead of a hardcoded value right now like in the example above?

**Comment by nonopolarity** [comment]
so I kind of get it... when you say imperative vs declarative... when it is done like https://codesandbox.io/s/wonderful-currying-9pheq  (the shaking text)
then it is the "declarative" way... and react-spring can do it that way too... but just a lot of re-render and if the component is already too CPU intensive it may not be too good to cause a lot of re-renders
